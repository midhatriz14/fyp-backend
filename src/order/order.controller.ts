import { Controller, Post, Body, Patch, Param } from "@nestjs/common";
import { OrderService } from "./order.service";

@Controller('orders')
export class OrderController {
    constructor(private readonly orderService: OrderService) { }

    @Post()
    async placeOrder(@Body() body: {
        organizerId: string;
        eventDate: string;
        eventTime: string;
        services: {
            vendorId: string;
            serviceName: string;
            price: number;
        }[];
    }) {
        try {
            // Call the service to create the order
            const order = await this.orderService.createOrder(
                body.organizerId,
                new Date(body.eventDate),
                body.eventTime,
                body.services
            );
            return order;
        } catch (error) {
            console.error('Error placing order:', error);
            throw new Error('Failed to place order');
        }
    }

    @Patch('vendor-response/:id')
    async respondToOrder(
        @Param('id') vendorOrderId: string,
        @Body() body: { status: 'accepted' | 'rejected'; message?: string },
    ) {
        return this.orderService.updateVendorResponse(vendorOrderId, body.status, body.message);
    }

    @Patch('complete-vendor/:id')
    async completeVendor(@Param('id') vendorOrderId: string) {
        return this.orderService.completeVendorOrder(vendorOrderId);
    }

    @Patch('complete-order/:id')
    async completeOrder(@Param('id') orderId: string) {
        return this.orderService.confirmOrderCompletion(orderId);
    }
}
