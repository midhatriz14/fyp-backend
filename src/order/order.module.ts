import { MongooseModule } from "@nestjs/mongoose";
import { Order, OrderSchema } from "../auth/schemas/order.schema";
import { VendorOrder, VendorOrderSchema } from "../auth/schemas/vendor-order.schema";
import { OrderController } from './order.controller';
import { OrderService } from './order.service';
import { Module } from '@nestjs/common';
import { User, UserSchema } from "src/auth/schemas/user.schema";

@Module({
    imports: [
        MongooseModule.forFeature([
            { name: Order.name, schema: OrderSchema },
            { name: User.name, schema: UserSchema },
            { name: VendorOrder.name, schema: VendorOrderSchema },
        ]),
    ],
    controllers: [OrderController],
    providers: [OrderService],
})
export class OrderModule { }
