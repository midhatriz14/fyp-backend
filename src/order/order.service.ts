import { Injectable, NotFoundException } from '@nestjs/common';
import { InjectModel } from '@nestjs/mongoose';
import { Model, Types } from 'mongoose';
import { Order } from 'src/auth/schemas/order.schema';
import { VendorOrder } from 'src/auth/schemas/vendor-order.schema';

@Injectable()
export class OrderService {
    constructor(
        @InjectModel(Order.name) private readonly orderModel: Model<Order>,
        @InjectModel(VendorOrder.name) private readonly vendorOrderModel: Model<VendorOrder>,
    ) { }

    // Create a new order
    async createOrder(
        organizerId: string,
        eventDate: Date,
        eventTime: string,
        services: { vendorId: string; serviceName: string; price: number }[],
        eventName: string,
        guests: number,
    ): Promise<Order> {
        // Calculate total and final amounts
        const totalAmount = services.reduce((sum, service) => sum + service.price, 0);
        const finalAmount = totalAmount; // Modify this if you want to apply discounts or other adjustments

        // Create the order document
        const order = new this.orderModel({
            organizerId: new Types.ObjectId(organizerId),
            eventDate,
            eventTime,
            eventName,
            guests,
            totalAmount,
            discount: 0, // You can modify this if needed
            finalAmount,
            status: 'pending', // Initial status
        });

        // Save the order document
        const savedOrder = await order.save();

        // Create vendor orders and push the vendorOrder _id to the order's vendorOrders field
        const vendorOrderIds: Types.ObjectId[] = [];
        for (const service of services) {
            const vendorOrder = new this.vendorOrderModel({
                orderId: savedOrder._id,
                vendorId: new Types.ObjectId(service.vendorId),
                serviceName: service.serviceName,
                price: service.price,
                status: 'pending', // Initial vendor order status
            });

            // Save each vendor order and push its _id into the vendorOrders array
            const savedVendorOrder = await vendorOrder.save();
            vendorOrderIds.push(savedVendorOrder._id);
        }

        // Update the order document with the vendorOrder IDs
        savedOrder.vendorOrders = vendorOrderIds;
        await savedOrder.save();

        return savedOrder;
    }


    // Get orders based on type (Organizer or Vendor)
    async getOrders(
        type: string,
        userId: string,
        status?: string,
        limit = 10,
        skip = 0,
    ): Promise<Order[]> {
        const query: any = {
            ...(status && { status }),  // Optional status filter
        };

        // Check type and modify query accordingly
        if (type === 'Organizer') {
            query.organizerId = userId;  // Filter orders by organizerId
        } else if (type === 'Vendor') {
            query.vendorOrders = { $elemMatch: { vendorId: userId } };  // Filter by vendorId in vendorOrders array
        }

        return this.orderModel
            .find(query)
            .skip(skip)
            .limit(limit)
            .populate('organizerId')  // Populate organizer details
            .populate({
                path: 'vendorOrders',  // Populate vendorOrders subdocument
                populate: {
                    path: 'vendorId',  // Populate vendor details inside each vendorOrder
                    model: 'User',  // Specify the 'Vendor' model to populate vendor info
                },
            })
            .exec();
    }




    // Get order stats (pending, processing, completed)
    async getOrderStats() {
        const totalOrders = await this.orderModel.countDocuments();
        const pending = await this.orderModel.countDocuments({ status: 'pending' });
        const processing = await this.orderModel.countDocuments({ status: 'processing' });
        const completed = await this.orderModel.countDocuments({ status: 'completed' });

        return {
            totalOrders,
            pending,
            processing,
            completed,
        };
    }

    // Update the status of an order to "completed"
    async completeOrder(orderId: string): Promise<Order> {
        const order = await this.orderModel.findById(orderId);

        if (!order) {
            throw new NotFoundException(`Order with ID ${orderId} not found`);
        }

        order.status = 'completed';
        return order.save();
    }

    // Update vendor order status (accepted/rejected)
    async updateVendorResponse(vendorOrderId: string, status: 'accepted' | 'rejected', message?: string) {
        const vendorOrder = await this.vendorOrderModel.findById(vendorOrderId);

        if (!vendorOrder) {
            throw new NotFoundException(`Vendor Order with ID ${vendorOrderId} not found`);
        }

        vendorOrder.status = status;
        if (message) {
            vendorOrder.message = message;
        }
        return vendorOrder.save();
    }

    // Mark a vendor order as completed
    async completeVendorOrder(vendorOrderId: string) {
        const vendorOrder = await this.vendorOrderModel.findById(vendorOrderId);

        if (!vendorOrder) {
            throw new NotFoundException(`Vendor Order with ID ${vendorOrderId} not found`);
        }

        vendorOrder.status = 'completed';
        return vendorOrder.save();
    }

    // Delete an order from the database
    async deleteOrder(orderId: string): Promise<any> {
        const order = await this.orderModel.findById(orderId);
        if (!order) {
            throw new NotFoundException(`Order with ID ${orderId} not found`);
        }

        // Delete associated vendor orders
        await this.vendorOrderModel.deleteMany({ orderId });

        return this.orderModel.deleteOne({ _id: orderId });
    }

    async confirmOrderCompletion(orderId: string) {
        return this.orderModel.findByIdAndUpdate(orderId, { status: 'completed' }, { new: true });
    }
}
