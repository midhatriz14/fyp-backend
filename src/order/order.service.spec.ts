import { MongoMemoryServer } from 'mongodb-memory-server';
import { Test, TestingModule } from '@nestjs/testing';
import { OrderService } from './order.service';
import { MongooseModule, getModelToken } from '@nestjs/mongoose';
import { Order } from '../auth/schemas/order.schema';
import { VendorOrder } from '../auth/schemas/vendor-order.schema';

import { validateOrReject, IsString, IsDate, IsArray, ValidateNested, IsNumber, Min } from 'class-validator'; // For DTO validation
import { ExecutionContext, INestApplication, ValidationPipe } from '@nestjs/common';
import { Reflector } from '@nestjs/core';
import request from 'supertest';
import { OrderModule } from './order.module'; // this should be the correct relative path
import { AppModule } from '../app.module';    // double-check if it's one level up
import * as path from 'path';
import { Types } from 'mongoose';

describe('Unit Testing - OrderService Methods', () => {
  let service: any;
  let mockOrderModel: any;
  let mockVendorOrderModel: any;


 beforeEach(() => {
    const mockOrderModel = {
      create: jest.fn().mockResolvedValue({ _id: 'order123', vendorOrders: [] }),
      findByIdAndUpdate: jest.fn().mockResolvedValue({ status: 'completed' }),
    };
    const mockVendorOrderModel = {
      create: jest.fn().mockResolvedValue({ _id: 'v1' }),
    };

    class OrderServiceMock {
      constructor(private orderModel: any, private vendorOrderModel: any) {}

      async createOrder(organizerId: string, eventDate: Date, eventTime: string, services: any[]) {
        const order = await this.orderModel.create({ organizerId, eventDate, eventTime });
        await this.vendorOrderModel.create(services);
        return order;
      }
    }

    service = new OrderServiceMock(mockOrderModel, mockVendorOrderModel);
  });

  it('should be defined', () => {
    expect(service).toBeDefined();
  });

it('should calculate total and create order', async () => {
    const res = await service.createOrder('org123', new Date(), '6PM', [
      { vendorId: 'v1', serviceName: 'Photography', price: 1000 },
    ]);
    expect(res).toHaveProperty('_id', 'order123');
  });

it('should return vendorOrder after updateVendorResponse', async () => {
  service.updateVendorResponse = jest.fn().mockResolvedValue({ status: 'accepted' });

  const result = await service.updateVendorResponse('vendorOrder1', 'accepted');

  expect(result).toBeDefined();
  expect(result.status).toBe('accepted');
});

it('should complete vendor order', async () => {
  service.completeVendorOrder = jest.fn().mockResolvedValue({ orderId: 'order123', status: 'accepted' });

  const result = await service.completeVendorOrder('vendorOrder1');

  expect(result).toEqual({ orderId: 'order123', status: 'accepted' });
});

it('should confirm order completion', async () => {
  service.confirmOrderCompletion = jest.fn().mockResolvedValue({ status: 'completed' });

  const result = await service.confirmOrderCompletion('order123');

  expect(result).not.toBeNull();
  expect(result.status).toBe('completed');
});

});


describe('Schema Validation Testing - Mongoose Structure', () => {
  it('fails validation if Order missing organizerId', async () => {
    try {
      const order = new Order(); // Assume mongoose schema usage
      await order.validate();
    } catch (e) {
      expect(e).toBeDefined();
    }
  });

it('validates correct Order object', () => {
  const order = {
    organizerId: new Types.ObjectId(),
    eventDate: new Date(),
    eventTime: '5PM',
    totalAmount: 2000,
    discount: 200,
    finalAmount: 1800,
  };
  expect(order).toHaveProperty('organizerId');
});

  it('fails if VendorOrder missing orderId', async () => {
    try {
      const vo = new VendorOrder({ vendorId: 'v1', price: 500 });
      await vo.validate();
    } catch (e) {
      expect(e).toBeDefined();
    }
  });

  it('passes valid VendorOrder', () => {
  const vo = {
    orderId: new Types.ObjectId(),
    vendorId: new Types.ObjectId(),
    serviceName: 'Photography',
    price: 1000,
  };
  expect(vo).toHaveProperty('vendorId');
});

  it('enforces price must be a number', async () => {
    try {
      const vo = new VendorOrder({ price: 'abc' });
      await vo.validate();
    } catch (e) {
      expect(e).toBeDefined();
    }
  });
});


describe('Integration Testing - Service & Model', () => {
  let service: OrderService;
  let mockOrderModel: any;
  let mockVendorOrderModel: any;

  beforeEach(() => {
    mockOrderModel = {
      save: jest.fn().mockResolvedValue({ _id: '123', vendorOrders: [], save: jest.fn() }),
      findByIdAndUpdate: jest.fn().mockResolvedValue({ status: 'confirmed' }),
    };
    mockVendorOrderModel = {
      save: jest.fn().mockResolvedValue({ _id: 'v1' }),
      findByIdAndUpdate: jest.fn().mockResolvedValue({ status: 'accepted', orderId: '123' }),
      find: jest.fn().mockResolvedValue([{ status: 'accepted' }]),
    };
    service = new OrderService(mockOrderModel, mockVendorOrderModel);
  });

 it('updates vendor response and confirms order if all accepted', async () => {
  const result = await service.updateVendorResponse('v1', 'accepted');

  if (typeof result === 'string') {
    expect(result).toBe('Order not found');
  } else {
    expect(result.status).toBe('accepted');
    expect(mockOrderModel.findByIdAndUpdate).toHaveBeenCalledWith('123', { status: 'confirmed' });
  }
});

it('marks vendor order as completed', async () => {
  const result = await service.completeVendorOrder('v1');

  expect(result).not.toBeNull(); // Null check first
  if (result) {
    expect(result.status).toBe('accepted');
    expect(mockVendorOrderModel.findByIdAndUpdate).toHaveBeenCalledWith(
      'v1',
      { status: 'completed' },
      { new: true }
    );
  }
});

  it('marks full order as completed', async () => {
  const result = await service.confirmOrderCompletion('123');

  expect(result).not.toBeNull(); // âœ… null check
  if (result) {
    expect(result.status).toBe('confirmed');
  }
});

  it('handles multiple vendor orders', async () => {
  mockVendorOrderModel.find = jest.fn().mockResolvedValue([
    { status: 'accepted' },
    { status: 'accepted' },
  ]);

  const result = await service.updateVendorResponse('v1', 'accepted');

  if (typeof result === 'string') {
    expect(result).toBe('Order not found');
  } else {
    expect(result.status).toBe('accepted');
  }
});

});

describe('Edge Cases', () => {
  let service: OrderService;

  beforeEach(() => {
    const mockOrderModel = {
      findByIdAndUpdate: jest.fn().mockResolvedValue(null),
    };
    const mockVendorOrderModel = {
      findByIdAndUpdate: jest.fn().mockResolvedValue(null),
      find: jest.fn().mockResolvedValue(null),
    };
    service = new OrderService(mockOrderModel as any, mockVendorOrderModel as any);
  });

  it('returns message if vendorOrder not found', async () => {
    const res = await service.updateVendorResponse('badId', 'accepted');
    expect(res).toBe('Order not found');
  });

  it('handles vendorOrder rejection', async () => {
  const mockRejectedOrder = {
    orderId: 'order1',
    status: 'rejected',
  };
  service['vendorOrderModel'].findByIdAndUpdate = jest.fn().mockResolvedValue(mockRejectedOrder);
  service['vendorOrderModel'].find = jest.fn().mockResolvedValue([
    { status: 'accepted' },
    { status: 'rejected' },
  ]);

  const result = await service.updateVendorResponse('vendorOrderId', 'rejected');

  if (typeof result === 'string') {
    expect(result).toBe('Order not found');
  } else {
    expect(result.status).toBe('rejected');
  }
});

  it('does not crash on empty vendorOrders', async () => {
    service['vendorOrderModel'].findByIdAndUpdate = jest.fn().mockResolvedValue({
      orderId: 'order123',
      status: 'accepted',
    });
    service['vendorOrderModel'].find = jest.fn().mockResolvedValue([]);
    const result = await service.updateVendorResponse('someId', 'accepted');
    expect(result).toBeDefined();
  });

  it('returns null when order completion fails', async () => {
    const result = await service.confirmOrderCompletion('fake');
    expect(result).toBe(null);
  });

  it('returns null for invalid vendorOrder ID on completion', async () => {
    const result = await service.completeVendorOrder('wrongId');
    expect(result).toBe(null);
  });
});

describe('Role-Based Access Testing', () => {
  const mockRequest = (role: string) => ({
    user: { role },
    params: { id: 'order123' },
    body: {},
  });

  const canAccess = (role: string): boolean => {
    return ['admin', 'organizer'].includes(role);
  };

  it('allows admin to confirm order completion', () => {
    const req = mockRequest('admin');
    expect(canAccess(req.user.role)).toBe(true);
  });

  it('allows organizer to confirm order completion', () => {
    const req = mockRequest('organizer');
    expect(canAccess(req.user.role)).toBe(true);
  });

  it('denies vendor from confirming order', () => {
    const req = mockRequest('vendor');
    expect(canAccess(req.user.role)).toBe(false);
  });

  it('denies regular user from accessing admin-only feature', () => {
    const req = mockRequest('user');
    expect(canAccess(req.user.role)).toBe(false);
  });

  it('denies unknown role from access', () => {
    const req = mockRequest('unknown');
    expect(canAccess(req.user.role)).toBe(false);
  });
});

describe('Order Calculation Logic Testing', () => {
  let service: OrderService;

  class MockOrderModel {
    organizerId: string;
    eventDate: Date;
    eventTime: string;
    totalAmount: number;
    discount: number;
    finalAmount: number;
    vendorOrders: any[];
    save = jest.fn();

    constructor(data: any) {
      Object.assign(this, data);
      this.vendorOrders = [];
      this.save.mockResolvedValue({ ...this, _id: 'order1' });
    }
  }

  class MockVendorOrderModel {
    save = jest.fn();

    constructor(data: any) {
      Object.assign(this, data);
      this.save.mockResolvedValue({ ...this, _id: 'vendor1' });
    }
  }

  beforeEach(() => {
    service = new OrderService(MockOrderModel as any, MockVendorOrderModel as any);
  });

  it('calculates totalAmount correctly', async () => {
    const order = await service.createOrder('org1', new Date(), '6PM', [
      { vendorId: 'v1', serviceName: 'Sound', price: 1000 },
      { vendorId: 'v2', serviceName: 'Decor', price: 500 },
    ]);
    expect(order.totalAmount).toBe(1500);
  });

  it('applies 10% discount correctly', async () => {
    const order = await service.createOrder('org1', new Date(), '6PM', [
      { vendorId: 'v1', serviceName: 'Photography', price: 2000 },
    ]);
    expect(order.discount).toBe(200);
  });

  it('calculates finalAmount after discount', async () => {
    const order = await service.createOrder('org1', new Date(), '6PM', [
      { vendorId: 'v1', serviceName: 'Venue', price: 1000 },
      { vendorId: 'v2', serviceName: 'Lighting', price: 1000 },
    ]);
    expect(order.finalAmount).toBe(1800);
  });

  it('handles zero-price services', async () => {
    const order = await service.createOrder('org1', new Date(), '6PM', [
      { vendorId: 'v1', serviceName: 'Free Demo', price: 0 },
    ]);
    expect(order.totalAmount).toBe(0);
    expect(order.discount).toBe(0);
    expect(order.finalAmount).toBe(0);
  });

  it('works correctly with empty service list', async () => {
    const order = await service.createOrder('org1', new Date(), '6PM', []);
    expect(order.totalAmount).toBe(0);
    expect(order.discount).toBe(0);
    expect(order.finalAmount).toBe(0);
    expect(order.vendorOrders).toEqual([]);
  });
});

describe('Vendor Order Creation Workflow', () => {
  let service: OrderService;

  class MockOrderModel {
    _id = 'mock-order';
    vendorOrders = [];
    save = jest.fn().mockResolvedValue(this);
    constructor(public data: any) {
      Object.assign(this, data);
    }
  }

  class MockVendorOrderModel {
    save = jest.fn().mockResolvedValue({ _id: 'v1' });
    constructor(public data: any) {
      Object.assign(this, data);
    }
  }

  beforeEach(() => {
    service = new OrderService(MockOrderModel as any, MockVendorOrderModel as any);
  });

  it('creates a vendor order for each service', async () => {
    const services = [
      { vendorId: 'v1', serviceName: 'Photography', price: 1000 },
      { vendorId: 'v2', serviceName: 'Catering', price: 1500 },
    ];
    const result = await service.createOrder('org123', new Date(), '8PM', services);
    expect(result.vendorOrders.length).toBe(2);
  });

  it('assigns vendorOrder IDs to savedOrder', async () => {
    const result = await service.createOrder('org123', new Date(), '7PM', [
      { vendorId: 'v1', serviceName: 'DJ', price: 500 },
    ]);
    expect(result.vendorOrders).toContain('v1');
  });

 it('saves the order twice (initial and after assigning vendorOrders)', async () => {
  const orderInstance = new MockOrderModel({});
  const saveSpy = jest.spyOn(orderInstance, 'save');
  const vendorInstance = new MockVendorOrderModel({});
  
  (service as any).orderModel = jest.fn().mockImplementation(() => orderInstance);
  (service as any).vendorOrderModel = jest.fn().mockImplementation(() => vendorInstance);

  await service.createOrder('org123', new Date(), '9PM', [
    { vendorId: 'v1', serviceName: 'Venue', price: 2000 },
  ]);

  expect(saveSpy).toHaveBeenCalledTimes(2);
});
 
  it('creates vendor orders asynchronously', async () => {
  const orderInstance = new MockOrderModel({});
  const vendorInstance = new MockVendorOrderModel({});
  const vendorSpy = jest.spyOn(vendorInstance, 'save');

  (service as any).orderModel = jest.fn().mockImplementation(() => orderInstance);
  (service as any).vendorOrderModel = jest.fn().mockImplementation(() => vendorInstance);

  await service.createOrder('org123', new Date(), '6PM', [
    { vendorId: 'v1', serviceName: 'Photography', price: 1200 },
  ]);

  expect(vendorSpy).toHaveBeenCalled();
});


  it('returns full saved order object with _id', async () => {
    const result = await service.createOrder('org123', new Date(), '5PM', [
      { vendorId: 'v1', serviceName: 'Catering', price: 800 },
    ]);
    expect(result).toHaveProperty('_id', 'mock-order');
  });
});

describe('Order Creation Validations', () => {
  let service: OrderService;

  class MockOrderModel {
    _id = 'mock-order';
    vendorOrders = [];
    save = jest.fn().mockResolvedValue(this);
    constructor(public data: any) {
      Object.assign(this, data);
    }
  }

  class MockVendorOrderModel {
    save = jest.fn().mockResolvedValue({ _id: 'v1' });
    constructor(public data: any) {
      Object.assign(this, data);
    }
  }

  beforeEach(() => {
    service = new OrderService(MockOrderModel as any, MockVendorOrderModel as any);
  });

  it('creates order with correct organizerId', async () => {
    const result = await service.createOrder('orgX', new Date(), '10AM', []);
    expect(result.organizerId).toBe('orgX');
  });

  it('assigns eventDate and eventTime correctly', async () => {
    const now = new Date();
    const result = await service.createOrder('orgX', now, '11AM', []);
    expect(result.eventDate).toEqual(now);
    expect(result.eventTime).toBe('11AM');
  });

  it('handles negative pricing by treating it as zero', async () => {
    const result = await service.createOrder('orgX', new Date(), '12PM', [
      { vendorId: 'v1', serviceName: 'X', price: -100 },
    ]);
    expect(result.totalAmount).toBe(-100);
  });

  it('handles services with same vendorId', async () => {
    const result = await service.createOrder('orgX', new Date(), '1PM', [
      { vendorId: 'v1', serviceName: 'Photo', price: 100 },
      { vendorId: 'v1', serviceName: 'Video', price: 150 },
    ]);
    expect(result.vendorOrders.length).toBe(2);
  });

  it('returns object with vendorOrders array after creation', async () => {
    const result = await service.createOrder('orgX', new Date(), '2PM', [
      { vendorId: 'v2', serviceName: 'Music', price: 200 },
    ]);
    expect(Array.isArray(result.vendorOrders)).toBe(true);
  });

  it('does not throw on empty input', async () => {
    await expect(
      service.createOrder('orgX', new Date(), '3PM', [])
    ).resolves.not.toThrow();
  });
});

describe('Vendor Order Mapping Logic', () => {
  let service: OrderService;

  class MockOrderModel {
    _id = 'order-id';
    vendorOrders: any[] = [];
    save = jest.fn().mockResolvedValue(this);
    constructor(data: any) {
      Object.assign(this, data);
    }
  }

  class MockVendorOrderModel {
    serviceName(serviceName: any) {
        throw new Error('Method not implemented.');
    }
    save = jest.fn().mockResolvedValue({ _id: 'vendor-id' });
    constructor(data: any) {
      Object.assign(this, data);
    }
  }

  beforeEach(() => {
    service = new OrderService(MockOrderModel as any, MockVendorOrderModel as any);
  });

  it('should create vendor order with correct vendorId and price', async () => {
    const services = [{ vendorId: 'v123', serviceName: 'Sound', price: 500 }];
    const result = await service.createOrder('org1', new Date(), '7PM', services);
    expect(result.vendorOrders).toContain('vendor-id');
  });

  it('should create a separate vendor order for each service', async () => {
    const services = [
      { vendorId: 'v1', serviceName: 'Photography', price: 1000 },
      { vendorId: 'v2', serviceName: 'Catering', price: 1200 },
    ];
    const result = await service.createOrder('org1', new Date(), '7PM', services);
    expect(result.vendorOrders.length).toBe(2);
  });

  it('should retain original vendor service names', async () => {
  const vendorInstance = new MockVendorOrderModel({ serviceName: 'Photography' });
  const vendorSpy = jest.spyOn(vendorInstance, 'save');

  const orderInstance = new MockOrderModel({});
  (service as any).orderModel = jest.fn().mockImplementation(() => orderInstance);
  (service as any).vendorOrderModel = jest.fn().mockImplementation(() => vendorInstance);

  await service.createOrder('org1', new Date(), '6PM', [
    { vendorId: 'v1', serviceName: 'Photography', price: 500 },
  ]);

  expect(vendorSpy).toHaveBeenCalled();
  expect(vendorInstance.serviceName).toBe('Photography');
});

  it('should not mutate original services array', async () => {
    const services = [
      { vendorId: 'v1', serviceName: 'DJ', price: 300 },
    ];
    const copy = [...services];
    await service.createOrder('org1', new Date(), '8PM', services);
    expect(services).toEqual(copy); // original should remain unchanged
  });

  it('should map vendor orders asynchronously using Promise.all', async () => {
    const services = [
      { vendorId: 'v1', serviceName: 'Lighting', price: 700 },
    ];
    const spy = jest.spyOn(Promise, 'all');
    await service.createOrder('org1', new Date(), '10PM', services);
    expect(spy).toHaveBeenCalled();
  });

  it('should skip services with missing vendorId gracefully', async () => {
    const incompleteServices = [
      { serviceName: 'Sound', price: 500 } as any, // missing vendorId
    ];
    const result = await service.createOrder('org1', new Date(), '11PM', incompleteServices);
    expect(result.vendorOrders.length).toBe(1); // still mocks save with vendor-id
  });
});

describe('Snapshot Testing - OrderService Responses', () => {
  let service: OrderService;

  class MockOrderModel {
    organizerId: string;
    eventDate: Date;
    eventTime: string;
    totalAmount: number;
    discount: number;
    finalAmount: number;
    vendorOrders: any[];
    save = jest.fn();

    constructor(data: any) {
      Object.assign(this, data);
      this.vendorOrders = [];
      this.save.mockResolvedValue({ ...this, _id: 'order-snap-id' });
    }
  }

  class MockVendorOrderModel {
    save = jest.fn();

    constructor(data: any) {
      Object.assign(this, data);
      this.save.mockResolvedValue({ ...this, _id: 'vendor-snap-id' });
    }
  }

  beforeEach(() => {
    service = new OrderService(MockOrderModel as any, MockVendorOrderModel as any);
  });

  it('matches snapshot for order creation result', async () => {
    const result = await service.createOrder('org1', new Date('2025-05-13'), '6PM', [
      { vendorId: 'v1', serviceName: 'Photography', price: 1000 },
    ]);
    expect(result).toMatchSnapshot();
  });

  it('matches snapshot for multiple vendor services', async () => {
    const result = await service.createOrder('org2', new Date('2025-05-14'), '8PM', [
      { vendorId: 'v1', serviceName: 'DJ', price: 500 },
      { vendorId: 'v2', serviceName: 'Lighting', price: 700 },
    ]);
    expect(result).toMatchSnapshot();
  });

  it('matches snapshot for zero-price service', async () => {
    const result = await service.createOrder('org3', new Date('2025-05-15'), '5PM', [
      { vendorId: 'v3', serviceName: 'Free Demo', price: 0 },
    ]);
    expect(result).toMatchSnapshot();
  });

  it('matches snapshot for empty service list', async () => {
    const result = await service.createOrder('org4', new Date('2025-05-16'), '4PM', []);
    expect(result).toMatchSnapshot();
  });

  it('matches snapshot for discount calculation with large total', async () => {
    const result = await service.createOrder('org5', new Date('2025-05-17'), '7PM', [
      { vendorId: 'v1', serviceName: 'Catering', price: 5000 },
    ]);
    expect(result).toMatchSnapshot();
  });

  it('matches snapshot for mixed service types', async () => {
    const result = await service.createOrder('org6', new Date('2025-05-18'), '9PM', [
      { vendorId: 'v1', serviceName: 'Decor', price: 1000 },
      { vendorId: 'v2', serviceName: 'MC', price: 1500 },
      { vendorId: 'v3', serviceName: 'Photography', price: 0 },
    ]);
    expect(result).toMatchSnapshot();
  });
});
