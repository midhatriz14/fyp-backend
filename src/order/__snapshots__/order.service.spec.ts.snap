// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot Testing - OrderService Responses matches snapshot for discount calculation with large total 1`] = `
{
  "_id": "order-snap-id",
  "discount": 500,
  "eventDate": 2025-05-17T00:00:00.000Z,
  "eventTime": "7PM",
  "finalAmount": 4500,
  "organizerId": "org5",
  "save": [MockFunction] {
    "calls": [
      [],
      [],
    ],
    "results": [
      {
        "type": "return",
        "value": Promise {},
      },
      {
        "type": "return",
        "value": Promise {},
      },
    ],
  },
  "totalAmount": 5000,
  "vendorOrders": [
    "vendor-snap-id",
  ],
}
`;

exports[`Snapshot Testing - OrderService Responses matches snapshot for empty service list 1`] = `
{
  "_id": "order-snap-id",
  "discount": 0,
  "eventDate": 2025-05-16T00:00:00.000Z,
  "eventTime": "4PM",
  "finalAmount": 0,
  "organizerId": "org4",
  "save": [MockFunction] {
    "calls": [
      [],
      [],
    ],
    "results": [
      {
        "type": "return",
        "value": Promise {},
      },
      {
        "type": "return",
        "value": Promise {},
      },
    ],
  },
  "totalAmount": 0,
  "vendorOrders": [],
}
`;

exports[`Snapshot Testing - OrderService Responses matches snapshot for mixed service types 1`] = `
{
  "_id": "order-snap-id",
  "discount": 250,
  "eventDate": 2025-05-18T00:00:00.000Z,
  "eventTime": "9PM",
  "finalAmount": 2250,
  "organizerId": "org6",
  "save": [MockFunction] {
    "calls": [
      [],
      [],
    ],
    "results": [
      {
        "type": "return",
        "value": Promise {},
      },
      {
        "type": "return",
        "value": Promise {},
      },
    ],
  },
  "totalAmount": 2500,
  "vendorOrders": [
    "vendor-snap-id",
    "vendor-snap-id",
    "vendor-snap-id",
  ],
}
`;

exports[`Snapshot Testing - OrderService Responses matches snapshot for multiple vendor services 1`] = `
{
  "_id": "order-snap-id",
  "discount": 120,
  "eventDate": 2025-05-14T00:00:00.000Z,
  "eventTime": "8PM",
  "finalAmount": 1080,
  "organizerId": "org2",
  "save": [MockFunction] {
    "calls": [
      [],
      [],
    ],
    "results": [
      {
        "type": "return",
        "value": Promise {},
      },
      {
        "type": "return",
        "value": Promise {},
      },
    ],
  },
  "totalAmount": 1200,
  "vendorOrders": [
    "vendor-snap-id",
    "vendor-snap-id",
  ],
}
`;

exports[`Snapshot Testing - OrderService Responses matches snapshot for order creation result 1`] = `
{
  "_id": "order-snap-id",
  "discount": 100,
  "eventDate": 2025-05-13T00:00:00.000Z,
  "eventTime": "6PM",
  "finalAmount": 900,
  "organizerId": "org1",
  "save": [MockFunction] {
    "calls": [
      [],
      [],
    ],
    "results": [
      {
        "type": "return",
        "value": Promise {},
      },
      {
        "type": "return",
        "value": Promise {},
      },
    ],
  },
  "totalAmount": 1000,
  "vendorOrders": [
    "vendor-snap-id",
  ],
}
`;

exports[`Snapshot Testing - OrderService Responses matches snapshot for zero-price service 1`] = `
{
  "_id": "order-snap-id",
  "discount": 0,
  "eventDate": 2025-05-15T00:00:00.000Z,
  "eventTime": "5PM",
  "finalAmount": 0,
  "organizerId": "org3",
  "save": [MockFunction] {
    "calls": [
      [],
      [],
    ],
    "results": [
      {
        "type": "return",
        "value": Promise {},
      },
      {
        "type": "return",
        "value": Promise {},
      },
    ],
  },
  "totalAmount": 0,
  "vendorOrders": [
    "vendor-snap-id",
  ],
}
`;
